src/esockd_acceptor.erl

```diff
@@ -237,32 +265,102 @@ handle_accept_error(enotconn, _, _) ->
+    {keep_state, State, {next_event, internal, accept_and_close}};
+handle_socket_error(Reason, State, _StateName) when Reason =:= econnaborted ->
     {next_state, waiting, State, {next_event, internal, begin_waiting}};
-%% emfile: The per-process limit of open file descriptors has been reached.
-%% enfile: The system limit on the total number of open files has been reached.
-%% enfile: The system limit on the total number of open files has been reached.
-handle_socket_error(Reason, State) when Reason =:= emfile; Reason =:= enfile ->
-    error_logger:error_msg(
-        "Accept error on ~s: ~s",
-        [esockd:format(State#state.sockname), explain_posix(Reason)]
-    ),
-    {next_state, suspending, State, {state_timeout, 1000, begin_waiting}};
-handle_socket_error(Reason, State) ->
+handle_socket_error(Reason, State, suspending) when Reason =:= emfile; Reason =:= enfile ->
+    log_system_limit(State, Reason),
+    {keep_state, State, {next_event, internal, accept_and_close}};
+handle_socket_error(Reason, State, _StateName) when Reason =:= emfile; Reason =:= enfile ->
+    log_system_limit(State, Reason),
+    enter_suspending(State, ?SYS_LIMIT_SUSPEND_MS);
+handle_socket_error(Reason, State, _StateName) ->
     {stop, Reason, State}.
 
 explain_posix(emfile) ->
     "EMFILE (Too many open files)";
 explain_posix(enfile) ->
     "ENFILE (File table overflow)".
+
+log_system_limit(State, Reason) ->
+    logger:log(critical,
+               #{msg => "cannot_accept_more_connections",
+                 listener => esockd:format(State#state.sockname),
+                 cause => explain_posix(Reason)}).
+
+enter_suspending(State, Timeout) ->
+    Actions = [{next_event, internal, accept_and_close},
+               {state_timeout, Timeout, begin_waiting}],
+    {next_state, suspending, State, Actions}.
+
+inc_stats(#state{proto = Proto, listen_on = ListenOn}, Tag) ->
+    Counter = counter(Tag),
+    case Counter of
+        closed_sys_limit ->
+            %% slow down when system limit reached
+            %% to aovid flooding the error logs
+            %% also, it makes no sense to drain backlog in such condition
+            timer:sleep(100);
+        _ ->
+            ok
+    end,
+    _ = esockd_server:inc_stats({Proto, ListenOn}, Counter, 1),
+    ok.
+
+counter(accepted) -> accepted;
+counter(emfile) -> closed_sys_limit;
+counter(enfile) -> closed_sys_limit;
+counter(overloaded) -> closed_overloaded;
+counter(rate_limited) -> closed_rate_limited;
+counter(_) -> closed_other_reasons.
+
+start_connection(ConnSup, Sock, UpgradeFuns) when is_pid(ConnSup) ->
+    esockd_connection_sup:start_connection(ConnSup, Sock, UpgradeFuns);
+start_connection({F, A}, Sock, UpgradeFuns) when is_function(F) ->
+    %% only in tests so far
+    apply(F, A ++ [Sock, UpgradeFuns]).
+
+%% To throttle system-limit error logs,
+%% if system limit reached, slow down the state machine by
+%% trunning the immediate return to a delayed async message.
```

Comment:
```suggestion
%% delaying the async return message for longer.
```
?

Reply:
it's not a async return, rather an immediate return from `prim_inet:async_accept`.

Reply:
Then:
```suggestion
%% turning the immediate return to a delayed async message
```
?